### PR TITLE
feat: add go plus api check

### PR DESCRIPTION
**Pull request by BillySh1**

### CODE REVIEW

File: scripts/contracts-risk-check.js [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+        res += num;
+      }
+    });
+  });
+  return res;
+};
+
+const singleCheck = async (chainId, address) => {
+  if (!supportChains.includes(chainId)) return;
+  if (safeChainIds.includes(chainId)) return;
```
Comment (guanbinrui) [inline scripts/contracts-risk-check.js:85]: Do you mean not include?

File: scripts/contracts-risk-check.js [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+//       riskRate: 0,
+//       riskDetail: null,
+//     },
+//   ],
+// };
+
+const getSupportedChains = async () => {
+  if (supportChains.length) return;
+  try {
+    await fetch("https://api.gopluslabs.io/api/v1/supported_chains")
```
Comment (guanbinrui) [inline scripts/contracts-risk-check.js:34]: A utility function for this and handle error in that function.

Comment (guanbinrui) [inline scripts/contracts-risk-check.js:34]: A utility function for this and handle error in that function.

File: scripts/contracts-risk-check.js [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        if (result.length) {
+          result.map((item) => [supportChains.push(Number(item.id))]);
+        }
+      });
+  } catch (e) {
+    return;
+  }
+};
+
+// const sleep = (ms) => {
```
Comment (guanbinrui) [inline scripts/contracts-risk-check.js:46]: Remove unused code.

File: scripts/contracts-risk-check.js [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+//   return new Promise((resolve) => setTimeout(resolve, ms));
+// };
+
+const riskCheck = async (list, rules) => {
+  await getSupportedChains();
+
+  const rule = rules ?? defaultRules;
+  const chainId = list.tokens[0].chainId;
+  const temp = JSON.parse(JSON.stringify(list));
```
Comment (guanbinrui) [inline scripts/contracts-risk-check.js:55]: Why you need to do this?

File: scripts/contracts-risk-check.js [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    rule[index].forEach((item) => {
+      if (keys.includes(item) && detail[item] === "1") {
+        res += num;
+      }
+    });
+  });
+  return res;
+};
+
+const singleCheck = async (chainId, address) => {
```
Comment (guanbinrui) [inline scripts/contracts-risk-check.js:83]: This API supports to handle more than one address in a single HTTP request.

File: scripts/generate-erc20.js [hunk 6]
```diff
@@ -149,7 +150,8 @@ const start = async () => {
     "^[a-zA-Z0-9+\\-%/\\$\\.]+$";
   const validate = ajv.compile(schema);
   if (validate(MaskTokenList)) {
-    process.stdout.write(JSON.stringify(MaskTokenList));
+    const riskCheckedList = await riskCheck(MaskTokenList);
+    process.stdout.write(JSON.stringify(riskCheckedList));
```
Comment (guanbinrui) [inline scripts/generate-erc20.js:154]: Dump the risk data into the final token list is unnecessary. Instead write into an audit file that let the reviewer to check the file.

Comment (BillySh1) [reply scripts/generate-erc20.js]: modified

File: scripts/risk-check.js [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+const Mumbai = require("../src/erc20/mumbai.json");
+const Aurora = require("../src/erc20/aurora.json");
+const Avalanche = require("../src/erc20/avalanche.json");
+const Boba = require("../src/erc20/boba.json");
+
+const GO_PLUS_LABS_ROOT_URL = "https://api.gopluslabs.io";
+const GO_PLUS_TOKEN_SECURITY_URL = "api/v1/token_security";
+const fetch = require("node-fetch");
+
+const chainId = Number.parseInt(process.argv.slice(2)[0]);
```
Comment (guanbinrui) [inline scripts/risk-check.js:22]: Plz give radix as the second parameter.

Comment (BillySh1) [reply scripts/risk-check.js:22]: ok

File: scripts/risk-check.js [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+  42220: Celo,
+  43114: Avalanche,
+  80001: Mumbai,
+  1313161554: Aurora,
+};
+
+const riskCheck = async () => {
+  try {
+    if (!supportedChainIds.includes(chainId)) {
+      process.stdout.write("unsupportedChainId");
```
Comment (guanbinrui) [inline scripts/risk-check.js:73]: `console.log`

File: scripts/risk-check.js [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+      return;
+    }
+    let res = null;
+    const addresses = chainIdToTokensMapping[chainId]
+      .map((item) => item.address)
+      .join(",");
+    const url = `${GO_PLUS_LABS_ROOT_URL}/${GO_PLUS_TOKEN_SECURITY_URL}/${chainId}?contract_addresses=${addresses}`;
+    await fetch(url)
```
Comment (guanbinrui) [inline scripts/risk-check.js:81]: `return fetch(url)`

File: scripts/risk-check.js [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+    process.stdout.write(JSON.stringify(res));
+    return res;
+  } catch (e) {
+    process.stdout.write(`risk check error for ChianId: ${chainId},error:`, e);
+  }
+};
+
+const caculateRiskRates = (item) => {
+  let res = 0;
+  rules.forEach((rule) => {
```
Comment (guanbinrui) [inline scripts/risk-check.js:117]: Use `rules.reduce` to refactor it.

File: scripts/risk-check.js [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+    let res = null;
+    const addresses = chainIdToTokensMapping[chainId]
+      .map((item) => item.address)
+      .join(",");
+    const url = `${GO_PLUS_LABS_ROOT_URL}/${GO_PLUS_TOKEN_SECURITY_URL}/${chainId}?contract_addresses=${addresses}`;
+    await fetch(url)
+      .then((r) => r.json())
+      .then((data) => {
+        res = data.result;
+        if (res) {
```
Comment (guanbinrui) [inline scripts/risk-check.js:85]: Early return to reduce nested spaces.

File: scripts/risk-check.js [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+const Mumbai = require("../src/erc20/mumbai.json");
+const Aurora = require("../src/erc20/aurora.json");
+const Avalanche = require("../src/erc20/avalanche.json");
+const Boba = require("../src/erc20/boba.json");
+
+const GO_PLUS_LABS_ROOT_URL = "https://api.gopluslabs.io";
+const GO_PLUS_TOKEN_SECURITY_URL = "api/v1/token_security";
+const fetch = require("node-fetch");
+
+const chainId = Number.parseInt(process.argv.slice(2)[0],10);
```
Comment (guanbinrui) [inline scripts/risk-check.js]:
```suggestion
const chainId = Number.parseInt(process.argv.slice(2)[0], 10);
```

File: scripts/risk-check.js (hunk 8 above)
Comment (BillySh1) [reply scripts/risk-check.js:73]: modified

File: scripts/risk-check.js (hunk 9 above)
Comment (BillySh1) [reply scripts/risk-check.js:81]: modified

File: scripts/risk-check.js (hunk 12 above)
Comment (BillySh1) [reply scripts/risk-check.js]: ok

File: scripts/risk-check.js (hunk 11 above)
Comment (BillySh1) [reply scripts/risk-check.js]: modified

File: scripts/risk-check.js [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+    const res = await fetch(url);
+    const data = await res.json()
+    if (data && data.result) {
+      const temp = data.result;
+      console.log(JSON.stringify(temp))
+      for (const key in temp) {
+        const item = temp[key];
+        caculateRiskRates(item, key);
+      }
+    }
```
Comment (guanbinrui) [inline scripts/risk-check.js:96]:
```suggestion
    if (data?.result) {
      const temp = data.result;
      for (const key in temp) {
        const item = temp[key];
        caculateRiskRates(item, key);
      }
    }
```

**Review by guanbinrui — APPROVED** [review]

**Pull request merged**